fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,367 @@
+/*
+ * TenantEntryCache.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(NO_INTELLISENSE) && !defined(FDBSERVER_TENANTENTRYCACHE_ACTOR_G_H)
+#define FDBSERVER_TENANTENTRYCACHE_ACTOR_G_H
+#include "fdbserver/TenantEntryCache.actor.g.h"
+#elif !defined(FDBSERVER_TENANTENTRYCACHE_ACTOR_H)
+#define FDBSERVER_TENANTENTRYCACHE_ACTOR_H
+
+#pragma once
+
+#include "fdbclient/DatabaseContext.h"
+#include "fdbclient/FDBOptions.g.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/RunTransaction.actor.h"
+#include "fdbclient/Tenant.h"
+#include "fdbclient/TenantManagement.actor.h"
+#include "fdbserver/Knobs.h"
+#include "flow/IndexedSet.h"
+
+#include <functional>
+#include <unordered_map>
+
+#include "flow/actorcompiler.h" // has to be last include
+
+#define INITIAL_CACHE_GEN 1
+
+using TenantEntryCacheGen = uint64_t;
+using TenantNameEntryPair = std::pair<TenantName, TenantMapEntry>;
+using TenantNameEntryPairVec = std::vector<TenantNameEntryPair>;
+
+enum class TenantEntryCacheRefreshReason { INIT = 1, PERIODIC_TASK = 2, CACHE_MISS = 3 };
+
+template <class T>
+struct TenantEntryCachePayload {
+	TenantName name;
+	TenantMapEntry entry;
+	TenantEntryCacheGen gen;
+	// Custom client payload
+	T payload;
+};
+
+template <class T>
+using TenantEntryCachePayloadFunc =
+    std::function<TenantEntryCachePayload<T>(const TenantName&, const TenantMapEntry&, const TenantEntryCacheGen)>;
+
+// In-memory cache for TenantEntryMap objects. It supports three indices:
+// 1. Lookup by 'TenantId'
+// 2. Lookup by 'TenantPrefix'
+// 3. Lookup by 'TenantName'
+//
+// TODO:
+// ----
+// The cache allows user to construct the 'cached object' by supplying a callback to allow constructing the cached
+// object. The cache implements a periodic refresh mechanism, polling underlying database for updates (add/remove
+// tenants), in future we might want to implement database range-watch to monitor such updates
+
+template <class T>
+class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyable {
```

Comment:
Idea is to incorporate keyrange-watches to handle updates/add/remove entries to TenantPrefix, the feature is being developed as a separate task (and for someother purpose)  and once done will be leverage here as well.  
For now, the cache is updated via a polling task.

Reply:
My main concern is that this could break if a tenant gets cycled. Trying to use that tenant for various purposes is going to repeatedly fail (often with a retryable error) until the refresh comes through if there is no way to provide feedback that the tenant is failing.

Reply:
Tenant getting cycled is dangerous for various reasons, especially given `tenant boundary == encryption domain boundary`. The code assumes: `tenantId, tenantPrefix and tenantName` are unique.

Reply:
These are not properties of tenants in general (they cannot be currently since we don't keep a record of historical tenant names). IDs and prefixes should be unique, but this cache isn't doing any ID based validation.

The way this works in the client cache is that if it uses an entry and submits its ID, the server will respond with an error if the ID doesn't match. Then the client would invalidate the cache entry and fetch the updated one with the new ID.

I don't think we can build the tenant cache with these assumptions, so we should plan for the possibility that a tenant entry is stale in this way.

Reply:
> (they cannot be currently since we don't keep a record of historical tenant names)

Seems like `byName` cache index is problematic, can't think of a better way then enumerating cached elements and searching for input name? However, ID and prefix you mentioned is unique, so should be fine. 

(Please correct me if am missing anything)

Reply:
There is nothing wrong with a name cache (this is what the client uses), but we just need to be able to invalidate the cache entry when a server responds with an error indicating that our entry was stale.

Reply:
Makes sense, I will extend the API to allow `removeByID/Prefix/Name` enabling clients to invalidate any specific entry. Thanks for pointing out.

Reply:
Added `removeEntry` APIs (by Id/prefix/name). 

Also updated the code to allow client to specify `cacheRefreshMode` ; for now `periodicRefresh` is the only mode supported, but, once keyrange-watch is implemented, we can extend the support for it as well. 

Thanks